### PR TITLE
Use `GitCommand` to git clone the repository

### PR DESCRIPTION
**Pull request by giordano**

### CODE REVIEW

File: src/AnalyzeRegistry.jl [hunk 1]
```diff
@@ -269,7 +271,9 @@ function analyze!(root, pkg::RegistryEntry; auth::GitHub.Authorization=github_au
         # or deleted them.  In these cases git would ask for username and
         # password, provide it with fake values just to move on:
         # https://stackoverflow.com/a/65705346/2442087
-        run(pipeline(`git clone -q --depth 1 --config credential.helper='!f() { echo -e "username=git\npassword="; }; f' $(repo) $(dest)`; stderr=devnull))
+        git() do git
+            run(pipeline(`$git clone -q --depth 1 --config credential.helper='!f() { echo -e "username=git\npassword="; }; f' $(repo) $(dest)`; stderr=Base.stderr))
```
Comment (giordano) [inline src/AnalyzeRegistry.jl:275]: `stderr` needs to be reverted to `devnull` before merging

**Comment by DilumAluthge** [comment]
You should probably set the Documentation job to clone the General registry over Git (instead of using the General registry from the Pkg server).

**Comment by giordano** [comment]
This needs to wait a bit.  `GitCommand` is thread-unsafe.  This function definition replaces `withenv` with `addenv`:
```julia
function git(;
             adjust_PATH::Bool = true,
             adjust_LIBPATH::Bool = true)
    git_path, env_mapping = _env_mapping(; adjust_PATH = adjust_PATH,
                                           adjust_LIBPATH = adjust_LIBPATH)
    return addenv(`$(git_path)`, env_mapping...)
end
```
but I still get some environment variables leaking on this PR, which isn't good.

**Comment by DilumAluthge** [comment]
If you end up getting this to work, is it worth upstreaming your safer `git` function to GitCommand?

**Comment by giordano** [comment]
Yes, testing here it's slightly easier (BTW, can we clone a different repository in the tests of `GitCommand`?  Cloning General takes forever for me)

**Comment by DilumAluthge** [comment]
> Yes, but testing here is slightly easier (BTW, can we clone a different repository in the tests of `GitCommand`? Cloning General takes forever for me)

Yeah definitely.

**Comment by giordano** [comment]
Well, tests are now passing also on Windows.... :confused:

**Comment by DilumAluthge** [comment]
🙈

**Comment by giordano** [comment]
Oh, cool, failing again.......

**Comment by giordano** [comment]
Now it fails on macOS with Julia v1.6 :sob: :sob: :sob:

**Comment by DilumAluthge** [comment]
Is it possible that something else (besides GitCommand) is modifying the environment?

**Comment by giordano** [comment]
I don't _see_ anything that should do that :shrug:

**Comment by DilumAluthge** [comment]
If this is due to GitCommand, it's really concerning, because it seems to be non-deterministic.

**Comment by DilumAluthge** [comment]
So, even though the tests passed on GitCommand, my guess is the bug is probably also there.

**Comment by DilumAluthge** [comment]
Can you stick a `@info "" orig_libpath` at the top of the test suite, so we can see what those values look like on the jobs that pass?

**Comment by giordano** [comment]
https://github.com/giordano/AnalyzeRegistry.jl/pull/39/checks?check_run_id=2162409689#step:6:134
```
   Evaluated: nothing == "/Users/runner/hostedtoolcache/julia/1.6.0-rc3/x64/lib/julia:/Users/runner/.julia/artifacts/28513525059d7ec39911c6cbeb8df1ff560ef837/lib:/Users/runner/.julia/artifacts/6b04ad7f71b1a29adad801f4fdd0d34b3535c638/lib:/Users/runner/.julia/artifacts/3ae59680bc7c4918874255e560d75a3361858e0d/lib:/Users/runner/.julia/artifacts/b979c48b40a24395485b4d0a15bc1db50b8d46a2/lib:/Users/runner/.julia/artifacts/ec7ccd15172f5e13c4fb3a9499b2e4eccb5c1830/lib:/Users/runner/hostedtoolcache/julia/1.6.0-rc3/x64/bin/../lib/julia:/Users/runner/hostedtoolcache/julia/1.6.0-rc3/x64/bin/../lib"
```
* `28513525059d7ec39911c6cbeb8df1ff560ef837` is [`Expat_jll`](https://github.com/JuliaBinaryWrappers/Expat_jll.jl/blob/2e74cd289a06b98c92f497564662bd564c36234d/Artifacts.toml#L74),
* `6b04ad7f71b1a29adad801f4fdd0d34b3535c638` is [`OpenSSL_jll`](https://github.com/JuliaBinaryWrappers/OpenSSL_jll.jl/blob/ba0879e50775d54bb75db544ba57e356093c0872/Artifacts.toml#L74`),
* `3ae59680bc7c4918874255e560d75a3361858e0d` is [`Libiconv_jll`](https://github.com/JuliaBinaryWrappers/Libiconv_jll.jl/blob/5335fc9772c03500063addbeb66f036cf8464853/Artifacts.toml#L74),
* `ec7ccd15172f5e13c4fb3a9499b2e4eccb5c1830` is [`Gettext_jll`](https://github.com/JuliaBinaryWrappers/Gettext_jll.jl/blob/69a80d27e847d2376fd73b89c832a0ef3c03713f/Artifacts.toml#L74)

These are all [dependencies of `Git_jll`](https://github.com/JuliaBinaryWrappers/Git_jll.jl/blob/e092c5c7f15585954967c5a96959115623f4b594/README.md#dependencies), so I'm 100% sure these variables are coming somehow from `Git_jll`, likely through `GitCommand`.

> it seems to be non-deterministic.

It is

**Comment by DilumAluthge** [comment]
https://github.com/JuliaBinaryWrappers/Git_jll.jl/blob/e092c5c7f15585954967c5a96959115623f4b594/src/wrappers/x86_64-linux-gnu.jl#L15

IIUC, this wrapper uses `withenv`: https://github.com/JuliaPackaging/JLLWrappers.jl/blob/1631db3e80c9d36d257adc222e766a22a73df8d1/src/runtime.jl#L37

Is `withenv` threadsafe?

We call the `Git_jll.git` function in GitCommand, for example https://github.com/JuliaVersionControl/GitCommand.jl/blob/64d6cff67b2fe3f7050d8108183139ecd3ab4f58/src/env.jl#L4 and https://github.com/JuliaVersionControl/GitCommand.jl/blob/64d6cff67b2fe3f7050d8108183139ecd3ab4f58/src/env.jl#L17

**Comment by giordano** [comment]
Two `git` functions are generated: https://github.com/JuliaPackaging/JLLWrappers.jl/blob/1631db3e80c9d36d257adc222e766a22a73df8d1/src/products/executable_generators.jl.  One to be called
```julia
git() do git
    run(`$git ...`)
end
```
which uses `withenv` and isn't thread-safe, and the other one is to be called
```julia
run(`$(git()) ...`)
```
which uses `addenv` which _**should**_ be thread-safe, as far as I understand.  `GitCommand` v2.2 should be using this second wrapper function.

This is why I did #42, which was also causing issues without `GitCommand`

**Comment by DilumAluthge** [comment]
But in https://github.com/JuliaVersionControl/GitCommand.jl/blob/master/src/env.jl, aren't we calling `Git_jll` with a do-block? So won't that call the old `withenv` version?

**Comment by DilumAluthge** [comment]
E.g. from https://github.com/JuliaVersionControl/GitCommand.jl/blob/64d6cff67b2fe3f7050d8108183139ecd3ab4f58/src/env.jl#L17

```
    git_path, env_mapping = Git_jll.git(; adjust_PATH = adjust_PATH,
                                          adjust_LIBPATH = adjust_LIBPATH) do git_path
```

**Comment by giordano** [comment]
Yeah, that's still calling the wrong method :disappointed:

**Comment by DilumAluthge** [comment]
Since the do-block will call the method which has `f::Function` as the first positional argument.

**Comment by DilumAluthge** [comment]
The problem is that, in GitCommand, I need to get the path to the `git` binary, because I use that path to locate a bunch of other stuff.

**Comment by giordano** [comment]
It's very late here, I don't have the time to look further into this, but that's likely the issue.  As far as I understand `addenv` is composable, so we should be able to do `addenv(Git_jll.git(), env_mapping...)` and get a new `Cmd` with the composed mappings

**Comment by giordano** [comment]
> The problem is that, in GitCommand, I need to get the path to the git binary, because I use that path to locate a bunch of other stuff.

That's always in `Git_jll.git_path`, it's a variable

**Comment by giordano** [comment]
And https://github.com/JuliaVersionControl/GitCommand.jl/blob/64d6cff67b2fe3f7050d8108183139ecd3ab4f58/src/env.jl#L21 can be replaced by
```julia
root = Git_jll.artifact_dir
```
so you don't really need the path of `git` anywhere

**Comment by DilumAluthge** [comment]
https://github.com/JuliaVersionControl/GitCommand.jl/pull/46

**Comment by DilumAluthge** [comment]
Update the compat entry to require GitCommand 2.2.1?

**Comment by DilumAluthge** [comment]
Obviously, even if CI is all green on this PR, that doesn't mean that we've fixed the bug, since it is nondeterministic.

But I do think that as of GitCommand 2.2.1, we have removed all uses of anything that calls `withenv`.

**Comment by giordano** [comment]
Ok, after https://github.com/JuliaVersionControl/GitCommand.jl/pull/46 I've never seen any failure in the tests here, after rerunning them several times.  Thanks a lot @DilumAluthge!  I think this is good to go now

**Review by ericphanson — APPROVED** [review]

**Pull request merged**